lib/Parse/ParseExpr.cpp

```diff
@@ -2959,38 +2959,57 @@ ExprResult Parser::ParseBoundsExpression() {
 }
 
 bool Parser::ParseRelativeBoundsClause(ExprResult &Expr) {
-  bool IsError = false;
-  RelativeBoundsClause *RelativeClause = nullptr;
-  ExprResult ConstExpr;
-  RangeBoundsExpr *Range = nullptr;
-
+  bool isError = false;
   IdentifierInfo *Ident = Tok.getIdentifierInfo();
   SourceLocation BoundsKWLoc = Tok.getLocation();
+  RangeBoundsExpr *Range = nullptr;
   Token TempTok = Tok;
   ConsumeToken();
-
   BalancedDelimiterTracker PT(*this, tok::l_paren);
-
   if (PT.expectAndConsume(diag::err_expected_lparen_after,
                           Ident->getNameStart())) {
     SkipUntil(tok::r_paren, StopAtSemi | StopBeforeMatch);
-    IsError = true;
-    return IsError;
+    isError = true;
+    return isError;
+  }
+
+  RelativeBoundsClause *RelativeClause =
+      CreateRelativeBoundsClause(isError, Ident, BoundsKWLoc);
+
+  if (Expr.isInvalid()) {
+    isError = true;
+    return isError;
+  }
+
+  if ((Range = dyn_cast<RangeBoundsExpr>(Expr.get()))) {
+    Range->setRelativeBoundsClause(RelativeClause);
+  } else {
+    Diag(TempTok, diag::err_expected_range_bounds_expr);
   }
 
+  PT.consumeClose();
+  return isError;
+}
+
+RelativeBoundsClause *
+Parser::CreateRelativeBoundsClause(bool &isError, IdentifierInfo *Ident,
```

Comment:
This should probably be called ParseRelativeBoundsClause, as this is what this method does.